ext/ffi/ffi.c

```diff
@@ -977,18 +1001,113 @@ static void zend_ffi_callback_trampoline(ffi_cif* cif, void* ret, void** args, v
 	free_alloca(fci.params, use_heap);
 
 	if (EG(exception)) {
+		// we're about to do a hard exit. unlock all mutexes
+		zend_ffi_dispatch_callback_end();
+		pthread_mutex_unlock(&FFI_G(vm_request_lock));
 		zend_error_noreturn(E_ERROR, "Throwing from FFI callbacks is not allowed");
 	}
 
 	ret_type = ZEND_FFI_TYPE(callback_data->type->func.ret_type);
 	if (ret_type->kind != ZEND_FFI_TYPE_VOID) {
-		zend_ffi_zval_to_cdata(ret, ret_type, &retval);
+		zend_ffi_zval_to_cdata(FFI_G(callback_data).ret, ret_type, &retval);
 	}
 
 	zval_ptr_dtor(&retval);
 }
 /* }}} */
 
+static void zend_ffi_interrupt_function(zend_execute_data *execute_data){ /* {{{ */
+	pthread_mutex_lock(&FFI_G(vm_request_lock));
```

Comment:
You should check if this is FFI related interrupt. This may be a POSIX signal or something else...